outrank/algorithms/synthetic_data_generators/cc_generator.py

```diff
@@ -0,0 +1,849 @@
+        :param N: Optional number of samples per class to downsample to
+        :param seed: Seed for random state of resample function
+        :param reshuffle: Reshuffle the dataset after downsampling
+        :return: Balanced X and y after downsampling
+        """
+
+        original_shape = X.shape
+
+        values, counts = np.unique(y, return_counts=True)
+        if N is None:
+            N = min(counts)
+
+        if N > min(counts):
+            raise ValueError('N must be equal to or less than the number of samples in minority class')
+
+        X_arrays_list = []
+        y_downsampled = []
+        for label in values:
+            X_label = [X[i] for i in range(len(y)) if y[i] == label]
+            X_label_downsample = resample(
+                X_label,
+                replace=True,
+                n_samples=N,
+                random_state=seed,
+            )
+            X_arrays_list.append(X_label_downsample)
+            ys = [label] * N
+            y_downsampled = np.concatenate((y_downsampled, ys), axis=0)
+
+        X_downsampled = np.concatenate(X_arrays_list, axis=0)
+
+        if reshuffle:
+            indices = np.arange(len(X_downsampled))
+            np.random.shuffle(indices)
+            X_downsampled = X_downsampled[indices]
+            y_downsampled = y_downsampled[indices]
+
+        downsampled_shape = X_downsampled.shape
+
+        self.dataset_info.update({
+            'downsampling': {
+                'original_shape': original_shape,
+                'downsampled_shape': downsampled_shape,
+            },
+        })
+
+        return X_downsampled, y_downsampled
+
+    def print_dataset(
+        self,
+        X: ArrayLike,
+        y: ArrayLike,
+    ):
+        """
+        Prints given dataset
+        :param X: dataset
+        :param y: labels
+        :return:
+        """
+
+        n_samples, n_features = X.shape
+        n = 0
+        for arr in X:
+            print('[', end='')
+            for i in range(n_features):
+                if i == n_features - 1:
+                    print(arr[i], end='')
+                else:
+                    print(arr[i], end=', ')
+            print(f'], Label: {y[n]}')
+            n += 1
+
+    def summarize(self):
+
+        print(f"Number of features: {self.dataset_info['general']['n_features']}")
+        print(f"Number of generated samples: {self.dataset_info['general']['n_samples']}")
+        if 'downsampling' in self.dataset_info:
+            print(
+                f"Dataset downsampled from shape {self.dataset_info['downsampling']['original_shape']}, to shape {self.dataset_info['downsampling']['downsampled_shape']}",
+            )
+        if 'n_class' in self.dataset_info['labels']:
+            print(f"Number of classes: {self.dataset_info['labels']['n_class']}")
+            print(f"Class relation: {self.dataset_info['labels']['class_relation']}")
+
+        print('-------------------------------------')
```

Comment:
Let's leave this for now, but outrank has proper logging which should be used when time comes, please put as todo

Reply:
Removed summarize function, added TODO